### PR TITLE
fix: Account Snap redirection dialog isn't displayed when sending transactions from MetaMask

### PR DESCRIPTION
**Pull request by owencraston**
## **Description**

Expected behaviour:
- GIVEN: user sends a transaction from a snap account that is using async approvals
- EXPECT: the `Finish signing` dialog should pop up and redirect the user to the companion dapp where they can complete the transaction

Actual behaviour:
- GIVEN: user sends a transaction from a snap account that is using async approvals
- ACTUAL: The transaction screen shows there is a pending approval and a small notification indicator is rendered in the fox icon. The user must then close and reopen MetaMask to see the finish signing dialog. 

### Why?
The issue is that the Home page in the extension checks for pending approvals and displays them on first load, but after that it won't redirect if an approval request is added and the UI is a popup. 

### What
This change will allow the Home popup to route to specific confirmations when it detects a pending approval of type `SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.showSnapAccountRedirect`. This is the approval type that we fire whenever a snap account is using `async approvals`. More approval types can easily be added in the future by changing the allow list values in `hasAllowedPopupRedirectApprovals`. **_This change will only effect flask for now_**. 

### How
In the Home component, receive a new prop that allows for confirmation redirects to be displayed. In the parent component we then allow list certain approval types (in our case `SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.showSnapAccountRedirect`). Instead of only rendering this approval when the user re opens MetaMask, we can dynamically route to the `Finish signing` dialog when it is called and the popup is still open.

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/157
This bug was introduced in this [PR](https://github.com/MetaMask/metamask-extension/pull/21312)

## **Manual testing steps**

1. run this branch locally and build using `yarn start:flask`
2. open the new dist in your browser
3. import an account that has funds. this can be testnet funds.
4. open https://metamask.github.io/snap-simple-keyring/latest/
5. click create account and then approval the account creation
6. turn off `Use Synchronous Approval
` in the Snap Simple Keyring dapp

Screenshot 2024-01-10 at 4 31 14 PM

7. you should see a new snap account in your Metamask

Screenshot 2024-01-10 at 4 29 18 PM

8. transfer a small amount of funds from your main account to this new snap account
9. once your snap account has funds, perform a send transaction from your snap account back to your main account
10. **_after you hit confirm on the new transaction, the Finish signing dialog should appear inside the pop up. This should be automatic._** 

Screenshot 2024-01-10 at 4 38 53 PM

11. Click go to site in the dialog, if should bring you back to the Snap Simple Keyring
12. click ` List Requests` in the snap site
13. copy the requestId from the json output
14. paste it into the `Approve request` field and click approve. 

### **Before**

https://github.com/MetaMask/metamask-extension/assets/22918444/98a5f8c0-e619-4b18-9b07-45f21d091bd2

### **After**

https://github.com/MetaMask/metamask-extension/assets/22918444/e89c1b93-dd2d-4585-9dd3-920877649322

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/background.js [hunk 1]
```diff
@@ -935,11 +935,28 @@ async function triggerUi() {
...
+  ];
+  const currentPendingApproval = Object.values(
+    controller.approvalController.state.pendingApprovals,
+  )[0];
+  if (
+    currentPendingApproval &&
+    FORCE_OPEN_APPROVALS.includes(currentPendingApproval?.type)
   ) {
+    shouldTriggerUi = true;
+  }
```
Comment (gantunesr) [inline app/scripts/background.js:956]: Just a nit, can we extract this into it's own method?

**Review by gantunesr — COMMENTED** [review]
The code seems like a reasonable fix. But it'd need a refactor in the future.

I did notice a bug where the request are not queuing while the confirmation screen is open. Juts reporting it here, not sure if it was caused by this PR.


https://github.com/MetaMask/metamask-extension/assets/17601467/b07bc26b-0a35-4a91-abb5-314dd2622161

File: app/scripts/background.js (hunk 1 above)
Comment (montelaidev) [reply app/scripts/background.js:956]: Just tested on develop and it looks like the issue is there as well.

**Comment by montelaidev** [comment]
Tested and the code looks ok

**Review by montelaidev — DISMISSED** [review]

**Review by gantunesr — APPROVED** [review]

**Review by HowardBraham — APPROVED** [review]

**Comment by k-g-j** [comment]
Worked as expected in manual testing on my end 
https://github.com/MetaMask/metamask-extension/assets/91970214/8e6b6e9d-93e0-43ac-82b2-1eb2a750d82c

**Comment by owencraston** [comment]
Confirmed that transactions/signs from a dapp are still working as expected.
https://github.com/MetaMask/metamask-extension/assets/22918444/3e57f87b-d1e5-4642-a537-2a859f95354c

**Pull request merged**